### PR TITLE
graph-api: added more validation

### LINKED ISSUE #496: graph-api endpoints data syntax verification
Checking type, length, format, ...

### PR DESCRIPTION
**Pull request by aramikm**
# Problem

Adds extra validation on graph-api dtos and request/path params

Closes #496 

# Discussion
- Since we are going to overhaul our way of e2e testing it didn't make sense to add tests which would be replaced in the future. 

## Steps to Verify:

1. Run the api
2. Go to swagger
3. Try some invalid values in requests

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -38,7 +38,9 @@
       "#graph-lib": ["libs/graph-lib/src"],
       "#graph-lib/*": ["libs/graph-lib/src/*"],
       "#types": ["libs/types/src"],
-      "#types/*": ["libs/types/src/*"]
+      "#types/*": ["libs/types/src/*"],
+      "#utils": ["libs/utils/src"],
```
Comment (aramikm) [inline tsconfig.json:42]: This is a new directory to store the common utils liek custom guards or decorators. please let me know if you have a better idea how to separate and keep these.

File: apps/graph-api/src/guards/read-only-deployment-guard.service.ts [hunk 2]
```diff
@@ -2,7 +2,7 @@ import { ConfigService } from '#graph-lib/config';
 import { Injectable, CanActivate, ExecutionContext } from '@nestjs/common';
 
 @Injectable()
-export class ReadOnlyGuard implements CanActivate {
+export class ReadOnlyDeploymentGuard implements CanActivate {
```
Comment (aramikm) [inline apps/graph-api/src/guards/read-only-deployment-guard.service.ts:5]: We have this Guard duplicated in other applications but I didn't want to increase the size of this PR but those will be also replaced and refactored in the future.

File: libs/types/src/dtos/graph/watch-graphs.dto.ts [hunk 3]
```diff
@@ -15,7 +16,7 @@ export class WatchGraphsDto {
 
   @IsNotEmpty()
   @IsString()
-  @IsUrl({ require_tld: false })
+  @IsUrl({ require_tld: false, require_protocol: true, require_valid_protocol: true })
```
Comment (saraswatpuneet) [inline libs/types/src/dtos/graph/watch-graphs.dto.ts:19]: These are good to know 💡

**Review by saraswatpuneet — APPROVED** [review]
👍🏽

File: apps/graph-api/src/guards/read-only-deployment-guard.service.ts (hunk 2 above)
Comment (saraswatpuneet) [reply apps/graph-api/src/guards/read-only-deployment-guard.service.ts:5]: Maybe we can add it to some existing story as a comment

File: apps/graph-api/src/controllers/v1/webhooks-v1.controller.ts [hunk 4]
```diff
@@ -51,8 +52,8 @@ export class WebhooksControllerV1 {
     required: false,
     description: "Boolean whether to include webhooks registered for 'all' MSA Ids (default: true)",
   })
-  async getWebhooksForMsa(@Param('msaId') msaId: string, @Query('includeAll') includeAll = 'true'): Promise<string[]> {
-    return this.apiService.getWebhooksForMsa(msaId, JSON.parse(includeAll ?? 'true'));
+  async getWebhooksForMsa(@Param() msaId: MsaIdParam, @Query('includeAll') includeAll = 'true'): Promise<string[]> {
+    return this.apiService.getWebhooksForMsa(msaId.msaId, JSON.parse(includeAll ?? 'true'));
```
Comment (JoeCap08055) [inline apps/graph-api/src/controllers/v1/webhooks-v1.controller.ts:56]:
```suggestion
  async getWebhooksForMsa(@Param() { msaId }: MsaIdParam, @Query('includeAll') includeAll = 'true'): Promise<string[]> {
    return this.apiService.getWebhooksForMsa(msaId, JSON.parse(includeAll ?? 'true'));
```

File: apps/graph-api/src/controllers/v1/webhooks-v1.controller.ts [hunk 5]
```diff
@@ -51,8 +52,8 @@ export class WebhooksControllerV1 {
     required: false,
     description: "Boolean whether to include webhooks registered for 'all' MSA Ids (default: true)",
   })
-  async getWebhooksForMsa(@Param('msaId') msaId: string, @Query('includeAll') includeAll = 'true'): Promise<string[]> {
-    return this.apiService.getWebhooksForMsa(msaId, JSON.parse(includeAll ?? 'true'));
+  async getWebhooksForMsa(@Param() msaId: MsaIdParam, @Query('includeAll') includeAll = 'true'): Promise<string[]> {
```
Comment (JoeCap08055) [inline apps/graph-api/src/controllers/v1/webhooks-v1.controller.ts:55]: I forget how we standardized parameter naming, but I _think_ we landed on suffixing REST input parameter object types with `Dto` or `DTO`.

Also, I would like to try and standardize on documenting the DTO classes with `@ApiProperty()` decorators, which eliminates the need to document each endpoint with `@ApiParameter()` or `@ApiQuery()`

File: apps/graph-api/src/controllers/v1/webhooks-v1.controller.ts [hunk 6]
```diff
@@ -65,8 +66,8 @@ export class WebhooksControllerV1 {
     example: 'http://localhost/webhook',
     description: 'URL for which to fetch all watched MSAs',
   })
-  async getWebhooksForUrl(@Query('url') url: string): Promise<string[]> {
-    return this.apiService.getWatchedGraphsForUrl(url);
+  async getWebhooksForUrl(@Query() url: UrlParam): Promise<string[]> {
+    return this.apiService.getWatchedGraphsForUrl(url.url);
```
Comment (JoeCap08055) [inline apps/graph-api/src/controllers/v1/webhooks-v1.controller.ts:70]:
```suggestion
  async getWebhooksForUrl(@Query() { url }: UrlParam): Promise<string[]> {
    return this.apiService.getWatchedGraphsForUrl(url);
```

File: apps/graph-api/src/controllers/v1/webhooks-v1.controller.ts [hunk 7]
```diff
@@ -106,8 +107,8 @@ export class WebhooksControllerV1 {
     example: '2',
     description: 'MSA for which to remove all webhook registrations',
   })
-  deleteWebhooksForMsa(@Param('msaId') msaId: string): Promise<void> {
-    return this.apiService.deleteWebhooksForUser(msaId);
+  deleteWebhooksForMsa(@Param() msaId: MsaIdParam): Promise<void> {
+    return this.apiService.deleteWebhooksForUser(msaId.msaId);
```
Comment (JoeCap08055) [inline apps/graph-api/src/controllers/v1/webhooks-v1.controller.ts:111]:
```suggestion
  deleteWebhooksForMsa(@Param() { msaId }: MsaIdParam): Promise<void> {
    return this.apiService.deleteWebhooksForUser(msaId);
```

File: apps/graph-api/src/controllers/v1/webhooks-v1.controller.ts [hunk 8]
```diff
@@ -120,7 +121,7 @@ export class WebhooksControllerV1 {
     example: 'http://localhost/webhook',
     description: 'URL for which to un-watch all MSAs',
   })
-  deleteAllWebhooksForUrl(@Query('url') url: string): Promise<void> {
-    return this.apiService.deleteWebhooksForUrl(url);
+  deleteAllWebhooksForUrl(@Query() url: UrlParam): Promise<void> {
+    return this.apiService.deleteWebhooksForUrl(url.url);
```
Comment (JoeCap08055) [inline apps/graph-api/src/controllers/v1/webhooks-v1.controller.ts:125]:
```suggestion
  deleteAllWebhooksForUrl(@Query() { url }: UrlParam): Promise<void> {
    return this.apiService.deleteWebhooksForUrl(url);
```

File: apps/graph-api/src/guards/read-only-deployment-guard.service.ts (hunk 2 above)
Comment (JoeCap08055) [reply apps/graph-api/src/guards/read-only-deployment-guard.service.ts:5]: We'll refactor this to `libs/utils`, but we can't until after we deal with the `ConfigService` dependency.

File: libs/types/src/dtos/common/params.dto.ts
Comment (JoeCap08055) [inline libs/types/src/dtos/common/params.dto.ts:1]: nit: rename classes to `xxxDto` to be consistent with the current naming pattern

File: libs/types/src/dtos/common/params.dto.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
+// eslint-disable-next-line max-classes-per-file
+import { ApiProperty } from '@nestjs/swagger';
+import { IsNotEmpty, IsNumberString, IsUrl } from 'class-validator';
+
+export class MsaIdParam {
+  @ApiProperty({ type: String, description: 'MSA Id of account' })
+  @IsNumberString({ no_symbols: true })
```
Comment (JoeCap08055) [inline libs/types/src/dtos/common/params.dto.ts:7]: I'd actually be in favor of creating a custom validation decorator, that can check not only for a valid numeric integer string, but also validate that it's > 0 (or any other simple validations on an MSA that don't rely on the chain)

File: libs/types/src/dtos/graph/graph-key-pair.dto.ts [hunk 10]
```diff
@@ -2,17 +2,24 @@ import { IsEnum, IsNotEmpty, IsString } from 'class-validator';
...
 import { HexString } from '@polkadot/util/types';
 import { ApiProperty } from '@nestjs/swagger';
+import { IsHexValue } from '#utils/decorators/is-hex-value.decorator';
 
 // DTO for the graph key pair
 export class GraphKeyPairDto {
-  @IsNotEmpty()
-  @IsString()
-  @ApiProperty({ description: 'Public graph encryption key as a hex string (prefixed with "0x")', type: String })
+  @IsHexValue({ minLength: 64, maxLength: 64, message: 'publicKey should be a 32 bytes value in hex format!' })
```
Comment (JoeCap08055) [inline libs/types/src/dtos/graph/graph-key-pair.dto.ts:9]: I wonder, is it worth it to write a decorator that's able to validate that a hex string decodes to a valid public key generated by a specific algorithm?

File: tsconfig.json (hunk 1 above)
Comment (JoeCap08055) [reply tsconfig.json:42]: You'll also need to add this path mapping to `jest.config.json`

**Review by mattheworris — APPROVED** [review]
- Read through the code
- Looks good, agree with other comments

🚢 it

File: tsconfig.json (hunk 1 above)
Comment (aramikm) [reply tsconfig.json:42]: Great catch!

File: libs/types/src/dtos/common/params.dto.ts (hunk 9 above)
Comment (aramikm) [reply libs/types/src/dtos/common/params.dto.ts:7]: I agree. I'll create customer MsaId and schemaId validators in here. 
On the side note `no_symbol` forces the numbers to be > 0

File: libs/types/src/dtos/graph/graph-key-pair.dto.ts (hunk 10 above)
Comment (aramikm) [reply libs/types/src/dtos/graph/graph-key-pair.dto.ts:9]: I have an issue for semantics checks and will add this inside that.

**Comment by JoeCap08055** [comment]
Regarding e2e tests: if we have a working e2e test framework, then it's worth while to add tests, because the test scenarios themselves for endpoint validation won't change--only the test framework setup will change, and we'll add additional tests for worker scenarios.

But if you want to add tests in a separate PR, that's fine.

**Comment by aramikm** [comment]
I think it would makes sense to add those tests when the framework is ready simply due to improved velocity of writing and verifying them compared to current state that if something goes wrong I would have to dig to see if my written test is the issue or the framework setup.

**Pull request merged**